### PR TITLE
Get projections just to work

### PR DESCRIPTION
**Pull request by woksin**

### CODE REVIEW

**Comment by woksin** [comment]
@jakhog This should be ready now. I'm just working on specs

File: Specifications/Events.Processing/Projections/for_ProjectionKeys/when_type_is_property.cs [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+            committed_event = committed_events.single("{\"property\": \"property\"}");
+            partition = Guid.Parse("8fe25727-b922-4907-aeaa-06497b0ba80e");
+            definition = given.projection_definition_builder.create()
+                            .with_selector(new ProjectionEventSelector(committed_event.Type, ProjectEventKeySelectorType.Property, "property"))
+                            .Build();
+        };
+        Because of = () => result = projection_keys.TryGetFor(definition, committed_event, partition, out key);
+
+        It should_get_key = () => result.ShouldBeTrue();
+        It should_have_the_correct_key = () => key.Value.ShouldEqual("property");
```
Comment (jakhog) [inline Specifications/Events.Processing/Projections/for_ProjectionKeys/when_type_is_property.cs:32]: Maybe give the property another value just to be sure we don't make any silly mistakes

File: Source/Events.Processing/Projections/EventProcessor.cs [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+            };
+            return await Process(@event, partitionId, request, cancellationToken).ConfigureAwait(false);
+        }
+
+        async Task<IProcessingResult> Process(CommittedEvent @event, PartitionId partitionId, ProjectionRequest request, CancellationToken token)
+        {
+            if (!_projectionKeys.TryGetFor(_projectionDefinition, @event, partitionId, out var projectionKey))
+            {
+                _logger.CouldNotGetProjectionKey(Identifier, Scope);
+                return new FailedProcessing("Could not get projection key");
```
Comment (jakhog) [inline Source/Events.Processing/Projections/EventProcessor.cs:83]: Would be nice to know what key and event failed here.

File: Source/Events.Processing/Projections/ProjectionsService.cs [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+        async Task ResetIfDefinitionChanged(
+            ProjectionRegistrationArguments arguments,
+            CancellationToken token)
+        {
+            _logger.ComparingProjectionDefintion(arguments);
+            var tenantsAndComparisonResult = await _projectionDefinitionComparer.DiffersFromPersisted(arguments.ProjectionDefinition, token).ConfigureAwait(false);
+
+            await _onAllTenants.PerformAsync(async tenant =>
+            {
+                if (!tenantsAndComparisonResult.TryGetValue(tenant, out var comparisonResult)) return;
```
Comment (jakhog) [inline Source/Events.Processing/Projections/ProjectionsService.cs:235]: It seems something terrible has gone wrong if the tenant is not in this dictionary?

File: Source/Events.Processing/Projections/ProjectionsService.cs [hunk 4]
```diff
@@ -0,0 +1,246 @@
...
+                _logger.PersistingProjectionDefinition(arguments, tenant);
+                if (!await _getProjectionDefinitions().TryPersist(arguments.ProjectionDefinition, token).ConfigureAwait(false)) throw new CouldNotPersistProjectionDefinition(arguments.ProjectionDefinition, tenant);
+                if (!comparisonResult.Succeeded)
+                {
+                    _logger.ResetingProjections(arguments, tenant, comparisonResult.FailureReason);
+                    if (!await _getProjectionStates().TryDrop(arguments.ProjectionDefinition.Projection, arguments.ProjectionDefinition.Scope, token).ConfigureAwait(false)) throw new CouldNotResetProjectionStates(arguments.ProjectionDefinition, tenant);
+                }
```
Comment (jakhog) [inline Source/Events.Processing/Projections/ProjectionsService.cs:242]: It seems we should do this in the opposite order? If the "new" definition is persisted - but the projections are not dropped. They will have a wrong old state - and not be replayed on next connection right?

**Review by jakhog — CHANGES_REQUESTED** [review]
Mostly small things. If there is one thing to take out and discuss is the support of "complex" property selector expressions in the keyselector. In my mind we should not support it until we decide on a standard way to do it - to not break it in the future.

File: Source/Events.Processing/Filters/FiltersProtocol.cs [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+    /// <typeparam name="TFilterRegistrationRequest"></typeparam>
+    /// <typeparam name="TFilterResponse"></typeparam>
+    /// <typeparam name="TRuntimeRegistrationArguments"></typeparam>
+    public interface IFiltersProtocol<TFilterClientMessage, TFilterRegistrationRequest, TFilterResponse, TRuntimeRegistrationArguments> : IReverseCallServiceProtocol<TFilterClientMessage, FilterRuntimeToClientMessage, TFilterRegistrationRequest, FilterRegistrationResponse, FilterEventRequest, TFilterResponse, TRuntimeRegistrationArguments>
+        where TFilterClientMessage : IMessage, new()
+        where TFilterRegistrationRequest : class
+        where TFilterResponse : class
+        where TRuntimeRegistrationArguments : class
+    {
+    }
```
[the remark this answers is not in the document]
Comment (woksin) [reply Source/Events.Processing/Filters/FiltersProtocol.cs:25]: yes 👍

File: Source/Events.Processing/Projections/CompareProjectionDefinitionsForAllTenants.cs [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+        {
+            var results = new Dictionary<TenantId, ProjectionDefinitionComparisonResult>();
+            await _onTenants.PerformAsync(async tenant =>
+            {
+                var definitions = _getDefinitions();
+                var tryGetDefinition = await definitions.TryGet(definition.Projection, definition.Scope, token).ConfigureAwait(false);
+                var comparisonResult = tryGetDefinition.Success switch
+                {
+                    true => CompareDefinitions(definition, tryGetDefinition),
+                    false => new ProjectionDefinitionComparisonResult()
```
[the remark this answers is not in the document]
Comment (woksin) [reply Source/Events.Processing/Projections/CompareProjectionDefinitionsForAllTenants.cs:49]: Could do that yes

File: Source/Events.Processing/Projections/CompareProjectionDefinitionsForAllTenants.cs [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+            if (CompareEvents(newDefinition, oldDefinition, out result)) return result;
+            return new ProjectionDefinitionComparisonResult();
+        }
+
+        bool CompareInitialState(ProjectionDefinition newDefinition, ProjectionDefinition oldDefinition, out ProjectionDefinitionComparisonResult result)
+        {
+            result = null;
+            if (newDefinition.InititalState != oldDefinition.InititalState)
+            {
+                result = new ProjectionDefinitionComparisonResult("The initial projection state is not the same as the persisted definition");
```
[the remark this answers is not in the document]
Comment (woksin) [reply Source/Events.Processing/Projections/CompareProjectionDefinitionsForAllTenants.cs:70]: Yes, that is nIce. We'll do that

File: Source/Events.Processing/Projections/CompareProjectionDefinitionsForAllTenants.cs [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+        bool CompareKeySelectorExpression(KeySelectorExpression newKeySelectorExpression, KeySelectorExpression oldKeySelectorExpression, out ProjectionDefinitionComparisonResult result)
+        {
+            result = null;
+            if (newKeySelectorExpression != oldKeySelectorExpression)
+            {
+                result = new ProjectionDefinitionComparisonResult("One or more key selector expressions does not match");
+                return true;
+            }
+            return false;
+        }
```
[the remark this answers is not in the document]
Comment (woksin) [reply Source/Events.Processing/Projections/CompareProjectionDefinitionsForAllTenants.cs:138]: It's to get nice error messages

File: Source/Events.Processing/Projections/CompareProjectionDefinitionsForAllTenants.cs [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+            return new ProjectionDefinitionComparisonResult();
+        }
+
+        bool CompareInitialState(ProjectionDefinition newDefinition, ProjectionDefinition oldDefinition, out ProjectionDefinitionComparisonResult result)
+        {
+            result = null;
+            if (newDefinition.InititalState != oldDefinition.InititalState)
+            {
+                result = new ProjectionDefinitionComparisonResult("The initial projection state is not the same as the persisted definition");
+                return true;
```
[the remark this answers is not in the document]
Comment (woksin) [reply Source/Events.Processing/Projections/CompareProjectionDefinitionsForAllTenants.cs:71]: Yes I agree

**Review by jakhog — APPROVED** [review]

**Pull request merged**